File: automation-test-selenium/src/test/java/org/suntower/fixtures/data/TestDataFactory.java
package org.suntower.fixtures.data;

import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import org.suntower.core.AppConfig;

public final class TestDataFactory {
  public static final int MISSING_ID = 999_999_999;
  public static final AuthPassword AUTH_PASSWORD = new AuthPassword();
  public static final AuthIdentity AUTH_IDENTITY = new AuthIdentity();
  public static final BuildingForm BUILDING_FORM = new BuildingForm();

  private static final AtomicInteger UNIQUE_COUNTER = new AtomicInteger();

  private TestDataFactory() {}

  public static String runToken() {
    String seed = System.getProperty("TEST_RUN_ID", System.getenv().getOrDefault("TEST_RUN_ID", "selenium-local"));
    String normalized = seed.toLowerCase(Locale.ROOT).replaceAll("[^a-z0-9]+", "");
    if (normalized.length() >= 8) {
      return normalized.substring(normalized.length() - 8);
    }
    return String.format("%8s", normalized).replace(' ', '0');
  }

  public static String uniqueCode(String prefix) {
    return "%s-%s-%s-%04d".formatted(
        prefix,
        runToken(),
        Long.toString(System.currentTimeMillis(), 36),
        UNIQUE_COUNTER.getAndIncrement() % 10_000);
  }

  public static String uniqueIdentifier(String prefix) {
    return uniqueCode(prefix).replaceAll("[^a-zA-Z0-9]", "");
  }

  public static String uniqueUsername(String prefix) {
    String value = uniqueIdentifier(prefix);
    return value.length() > 30 ? value.substring(0, 30) : value;
  }

  public static String uniqueEmail(String prefix) {
    return uniqueCode(prefix) + "@example.com";
  }

  public static String uniqueBuildingName(String prefix) {
    return prefix + " " + uniqueCode("building");
  }

  public static String uniquePhoneNumber() {
    long value = Math.abs(System.nanoTime() + UNIQUE_COUNTER.getAndIncrement());
    return "09" + String.format("%08d", value % 100_000_000);
  }

  public static CustomerPayload buildCustomerPayload() {
    String suffix = uniqueCode("customer");
    return new CustomerPayload(
        uniqueUsername("selcust"),
        AppConfig.get().defaultPassword(),
        "SEL Customer " + suffix,
        uniquePhoneNumber(),
        uniqueEmail("sel-customer"));
  }

  public static StaffPayload buildStaffPayload(String role) {
    String normalizedRole = role == null || role.isBlank() ? "STAFF" : role;
    String suffix = uniqueCode(normalizedRole.toLowerCase(Locale.ROOT));
    return new StaffPayload(
        uniqueUsername("selstf"),
        AppConfig.get().defaultPassword(),
        "SEL " + normalizedRole + " " + suffix,
        uniquePhoneNumber(),
        uniqueEmail("sel-staff"),
        normalizedRole);
  }

  public static Map<String, Object> buildBuildingPayload(String transactionType) {
    String normalizedTransactionType = transactionType == null || transactionType.isBlank() ? "FOR_RENT" : transactionType;
    String suffix = uniqueCode("building");
    boolean forRent = "FOR_RENT".equals(normalizedTransactionType);
    Map<String, Object> payload = new LinkedHashMap<>();
    payload.put("districtId", 1);
    payload.put("numberOfFloor", 10);
    payload.put("numberOfBasement", 1);
    payload.put("floorArea", 200);
    payload.put("rentPrice", forRent ? 1_000_000 : null);
    payload.put("deposit", forRent ? 2_000_000 : null);
    payload.put("serviceFee", forRent ? 100_000 : null);
    payload.put("carFee", forRent ? 50_000 : null);
    payload.put("motorbikeFee", forRent ? 20_000 : null);
    payload.put("waterFee", forRent ? 15_000 : null);
    payload.put("electricityFee", forRent ? 3_500 : null);
    payload.put("salePrice", forRent ? null : 3_000_000_000L);
    payload.put("name", "SEL Building " + suffix);
    payload.put("ward", "Xuan La");
    payload.put("street", "Vo Chi Cong");
    payload.put("propertyType", "OFFICE");
    payload.put("transactionType", normalizedTransactionType);
    payload.put("direction", "DONG");
    payload.put("level", "A");
    payload.put("taxCode", uniqueNumberCode("SEL", 10));
    payload.put("linkOfBuilding", "https://example.com");
    payload.put("image", null);
    payload.put("rentAreaValues", forRent ? "50,100" : "");
    payload.put("latitude", 21.0686);
    payload.put("longitude", 105.8033);
    payload.put("staffIds", java.util.List.of());
    return payload;
  }

  public static String uniqueNumberCode(String prefix, int digits) {
    String numeric = String.valueOf(System.currentTimeMillis()) + String.format("%06d", UNIQUE_COUNTER.getAndIncrement() % 1_000_000);
    String suffix = numeric.substring(Math.max(0, numeric.length() - digits));
    return prefix + suffix;
  }

  public record CustomerPayload(String username, String password, String fullName, String phone, String email) {}

  public record StaffPayload(String username, String password, String fullName, String phone, String email, String role) {}

  public static final class AuthPassword {
    public final String registrationDefault = "Password@123";
    public final String resetNewPassword = "Password@456";
    public final String shortOtp = "123456";
    public final String mismatchConfirmation = "Mismatch@456";
    public final String invalidLoginPassword = "wrong-password";
    public final String invalidRegisteredPassword = "WrongPassword!123";
  }

  public static final class AuthIdentity {
    public final String unknownUsername = "unknown-user";
  }

  public static final class BuildingForm {
    public final String taxCodePrefix = "TAX";
    public final String validLink = "https://example.com/building";
    public final String defaultDistrictId = "1";
    public final String defaultWard = "Xuan La";
    public final String defaultStreet = "Vo Chi Cong";
    public final String defaultLevel = "A";
    public final String defaultDirection = "DONG";
    public final Coordinates defaultCoordinates = new Coordinates(21.0686, 105.8033);
    public final RentalFields rentalCreate =
        new RentalFields(12, 2, 450, 1_200_000, 2_400_000, 100_000, 50_000, 20_000, 15_000, 3_500, "50,100");
    public final RentalFields rentalUpdate =
        new RentalFields(15, null, 999, 1_300_000, 2_500_000, 110_000, 60_000, 30_000, 18_000, 4_000, "70,140");
  }

  public record Coordinates(Number latitude, Number longitude) {}

  public record RentalFields(
      Number numberOfFloor,
      Number numberOfBasement,
      Number floorArea,
      Number rentPrice,
      Number deposit,
      Number serviceFee,
      Number carFee,
      Number motorbikeFee,
      Number waterFee,
      Number electricityFee,
      String rentAreaValues) {}
}

File: automation-test-playwright/fixtures/test-data/factories/TestDataFactory.ts
import { env } from "@helpers-runtime/env";
import { runtimePaths } from "@helpers-runtime/paths";

export class TestDataFactory {
  static readonly missingId = 999999999;
  static readonly missingSmallId = 999999;
  static readonly invoiceStatus = {
    pending: "PENDING",
    paid: "PAID",
    overdue: "OVERDUE"
  } as const;

  static readonly paymentMethod = {
    bankQr: "BANK_QR"
  } as const;

  static readonly transactionType = {
    rent: "FOR_RENT",
    sale: "FOR_SALE"
  } as const;

  static readonly testAmount = {
    staffInvoiceUpdateTotal: 9999,
    adminInvoiceUpdateTotal: 19999,
    adminInvoiceRejectedUpdateTotal: 99999
  } as const;

  static readonly testDate = {
    farPastDueDate: "2000-01-01"
  } as const;

  static readonly authPassword = {
    registrationDefault: "Password@123",
    resetNewPassword: "Password@456",
    shortOtp: "123456",
    mismatchConfirmation: "Mismatch@456",
    invalidLoginPassword: "wrong-password",
    invalidRegisteredPassword: "WrongPassword!123"
  } as const;

  static readonly authIdentity = {
    unknownUsername: "unknown-user"
  } as const;

  static readonly buildingForm = {
    taxCodePrefix: "TAX",
    validLink: "https://example.com/building",
    defaultDistrictId: String(env.testDataSeed.districtId),
    defaultWard: env.testDataSeed.ward,
    defaultStreet: env.testDataSeed.street,
    defaultLevel: "A",
    defaultDirection: "DONG",
    defaultCoordinates: {
      latitude: env.testDataSeed.latitude,
      longitude: env.testDataSeed.longitude
    },
    rentalCreate: {
      numberOfFloor: 12,
      numberOfBasement: 2,
      floorArea: 450,
      rentPrice: 1200000,
      deposit: 2400000,
      serviceFee: 100000,
      carFee: 50000,
      motorbikeFee: 20000,
      waterFee: 15000,
      electricityFee: 3500,
      rentAreaValues: "50,100"
    },
    rentalUpdate: {
      numberOfFloor: 15,
      floorArea: 999,
      rentPrice: 1300000,
      deposit: 2500000,
      serviceFee: 110000,
      carFee: 60000,
      motorbikeFee: 30000,
      waterFee: 18000,
      electricityFee: 4000,
      rentAreaValues: "70,140"
    }
  } as const;

  private static uniqueCounter = 0;

  static runToken(): string {
    return runtimePaths.runId
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, "")
      .slice(-8)
      .padStart(8, "0");
  }

  private static uniqueNumericString(length = 10): string {
    const timestamp = Date.now().toString();
    const counter = (this.uniqueCounter++ % 1000000).toString().padStart(6, "0");
    return `${timestamp}${counter}`.slice(-length);
  }

  private static baseDate(): Date {
    const configured = process.env.TEST_BASE_DATE;
    const parsed = configured ? new Date(configured) : new Date();
    return Number.isNaN(parsed.getTime()) ? new Date() : parsed;
  }

  private static formatDate(value: Date): string {
    return `${value.getFullYear()}-${String(value.getMonth() + 1).padStart(2, "0")}-${String(value.getDate()).padStart(2, "0")}`;
  }

  private static addMonths(value: Date, months: number): Date {
    const result = new Date(value);
    result.setMonth(result.getMonth() + months);
    return result;
  }

  static uniqueCode(prefix = "pw"): string {
    const timestamp = Date.now().toString(36);
    const counter = (this.uniqueCounter++ % 1679616).toString(36).padStart(4, "0");
    return `${prefix}-${this.runToken()}-${timestamp}-${counter}`;
  }

  static assertIncludesRunToken(value: string, label = "test data value"): void {
    if (!value.toLowerCase().includes(this.runToken())) {
      throw new Error(`${label} must include runToken ${this.runToken()} to keep cleanup safe and parallel-friendly.`);
    }
  }



  static uniqueIdentifier(prefix = "pw"): string {
    return this.uniqueCode(prefix).replace(/[^a-z0-9]/gi, "");
  }

  static uniqueUsername(prefix = "pw"): string {
    return this.uniqueIdentifier(prefix).slice(0, 30);
  }

  static uniqueBuildingName(prefix = "PW Building"): string {
    return `${prefix} ${this.uniqueCode("building")}`;
  }

  static uniqueCustomerName(prefix = "PW Customer"): string {
    return `${prefix} ${this.uniqueCode("customer")}`;
  }

  static uniqueEmail(prefix = "pw-user"): string {
    return `${this.uniqueCode(prefix)}@example.com`;
  }

  static uniquePhoneNumber(): string {
    return `0${this.uniqueNumericString(9)}`;
  }

  static uniqueNumberCode(prefix = "PW", digits = 10): string {
    const runDigits = this.runToken().replace(/\D/g, "").slice(-Math.min(4, digits));
    const remainingDigits = digits - runDigits.length;
    return `${prefix}${runDigits}${remainingDigits > 0 ? this.uniqueNumericString(remainingDigits) : ""}`;
  }

  static buildAdminStaffPayload(
    overrides: Record<string, unknown> = {},
    role: "STAFF" | "ADMIN" = "STAFF"
  ): Record<string, unknown> {
    const suffix = this.uniqueCode(role.toLowerCase());
    return {
      username: this.uniqueUsername(role.toLowerCase()),
      password: env.defaultPassword,
      fullName: `PW ${role} ${suffix}`,
      phone: this.uniquePhoneNumber(),
      email: this.uniqueEmail(`pw-${role.toLowerCase()}`),
      role,
      ...overrides
    };
  }

  static buildCustomerPayload(overrides: Record<string, unknown> = {}): Record<string, unknown> {
    const suffix = this.uniqueCode("customer");
    return {
      username: this.uniqueUsername("pwcust"),
      password: env.defaultPassword,
      fullName: `PW Customer ${suffix}`,
      phone: this.uniquePhoneNumber(),
      email: this.uniqueEmail("pw-customer"),
      staffIds: [],
      ...overrides
    };
  }

  static buildBuildingPayload(
    overrides: Record<string, unknown> = {},
    transactionType: "FOR_RENT" | "FOR_SALE" = TestDataFactory.transactionType.rent
  ): Record<string, unknown> {
    const suffix = this.uniqueCode("building");
    return {
      districtId: env.testDataSeed.districtId,
      numberOfFloor: 10,
      numberOfBasement: 1,
      floorArea: 200,
      rentPrice: transactionType === TestDataFactory.transactionType.rent ? 1000000 : null,
      deposit: transactionType === TestDataFactory.transactionType.rent ? 2000000 : null,
      serviceFee: transactionType === TestDataFactory.transactionType.rent ? 100000 : null,
      carFee: transactionType === TestDataFactory.transactionType.rent ? 50000 : null,
      motorbikeFee: transactionType === TestDataFactory.transactionType.rent ? 20000 : null,
      waterFee: transactionType === TestDataFactory.transactionType.rent ? 15000 : null,
      electricityFee: transactionType === TestDataFactory.transactionType.rent ? 3500 : null,
      salePrice: transactionType === TestDataFactory.transactionType.sale ? 3000000000 : null,
      name: `PW Building ${suffix}`,
      ward: env.testDataSeed.ward,
      street: env.testDataSeed.street,
      propertyType: "OFFICE",
      transactionType,
      direction: "DONG",
      level: "A",
      taxCode: this.uniqueNumberCode("PW", 10),
      linkOfBuilding: "https://example.com",
      image: null,
      rentAreaValues: "50,100",
      latitude: env.testDataSeed.latitude,
      longitude: env.testDataSeed.longitude,
      staffIds: [],
      ...overrides
    };
  }

  static buildContractPayload(overrides: Record<string, unknown> = {}): Record<string, unknown> {
    const baseDate = this.baseDate();
    const startDate = new Date(baseDate.getFullYear(), 0, 1);
    const endDate = new Date(baseDate.getFullYear() + 1, 11, 31);

    return {
      customerId: env.testDataSeed.customerId,
      buildingId: env.testDataSeed.buildingId,
      staffId: env.testDataSeed.staffId,
      rentPrice: 1000000,
      rentArea: 50,
      startDate: this.formatDate(startDate),
      endDate: this.formatDate(endDate),
      status: "ACTIVE",
      ...overrides
    };
  }

  static buildInvoicePayload(overrides: Record<string, unknown> = {}): Record<string, unknown> {
    const invoiceDate = this.baseDate();
    invoiceDate.setMonth(invoiceDate.getMonth() - 1);
    const month = invoiceDate.getMonth() + 1;
    const year = invoiceDate.getFullYear();
    const dueDateValue = new Date(year, month, 10);
    const dueDate = `${dueDateValue.getFullYear()}-${String(dueDateValue.getMonth() + 1).padStart(2, "0")}-10`;

    return {
      contractId: env.testDataSeed.contractId,
      customerId: env.testDataSeed.customerId,
      month,
      year,
      status: TestDataFactory.invoiceStatus.pending,
      dueDate,
      totalAmount: 1500000,
      details: [{ description: "Phi dich vu test", amount: 1500000 }],
      electricityUsage: 10,
      waterUsage: 5,
      ...overrides
    };
  }

  static buildSaleContractPayload(overrides: Record<string, unknown> = {}): Record<string, unknown> {
    return {
      buildingId: env.testDataSeed.buildingId,
      customerId: env.testDataSeed.customerId,
      staffId: env.testDataSeed.staffId,
      salePrice: 3000000000,
      transferDate: null,
      note: "Hop dong mua ban test",
      ...overrides
    };
  }

  static buildPropertyRequestPayload(
    overrides: Record<string, unknown> = {},
    requestType: "RENT" | "BUY" = "RENT"
  ): Record<string, unknown> {
    const desiredStartDate = this.addMonths(this.baseDate(), 2);
    const desiredEndDate = this.addMonths(desiredStartDate, 12);

    return {
      buildingId: env.testDataSeed.buildingId,
      requestType,
      desiredArea: requestType === "RENT" ? 80 : null,
      desiredStartDate: requestType === "RENT" ? this.formatDate(desiredStartDate) : null,
      desiredEndDate: requestType === "RENT" ? this.formatDate(desiredEndDate) : null,
      offeredPrice: requestType === "BUY" ? 3100000000 : 1200000,
      message: `Playwright property request ${this.uniqueCode("request")}`,
      ...overrides
    };
  }
}
